src/ui/lua/ScriptObject.mjs

```diff
@@ -0,0 +1,83 @@
+import {
+  LUA_REGISTRYINDEX,
+  lua_createtable,
+  lua_getglobal,
+  lua_pushlightuserdata,
+  lua_pushnumber,
+  lua_rawgeti,
+  lua_rawset,
+  lua_setglobal,
+  lua_setmetatable,
+  lua_settop,
+  lua_type,
+  luaL_ref,
+} from './api';
+
+const metaTables = new Map();
+const objectTypes = new Map();
+
+class LuaScriptObject {
+  constructor(ui) {
+    this.ui = ui;
+
+    this.luaRegistered = false;
+    this.luaRef = null;
+  }
+
+  register(name) {
+    const L = this.ui.lua.state;
+
+    if (!this.luaRegistered) {
+      // TODO: Tainted code handling
+
+      lua_createtable(L, 0, 0);
+      lua_pushnumber(L, 0.0);
+      lua_pushlightuserdata(L, this);
+      lua_rawset(L, -3);
+
+      const ref = this.constructor.getMetaTable(L);
+      lua_rawgeti(L, LUA_REGISTRYINDEX, ref);
+      lua_setmetatable(L, -2);
+
+      this.luaRef = luaL_ref(L, LUA_REGISTRYINDEX);
+    }
+
+    this.luaRegistered = true;
+
+    if (name) {
+      lua_getglobal(L, name);
+
+      const found = lua_type(L, -1);
+      lua_settop(L, -2);
+
+      if (!found) {
+        lua_rawgeti(L, LUA_REGISTRYINDEX, this.luaRef);
+        lua_setglobal(L, name);
+      }
+    }
+  }
+
+  deregister() {
+    // TODO: Unregister
+  }
+
+  static getMetaTable(L) {
+    let metaTable = metaTables.get(this);
+    if (!metaTable) {
+      metaTable = L.client.ui.lua.createMetaTable(this.scriptFunctions);
```

Comment:
This is pretty confusing, I think. Why does `L` have access to `client.ui`? To follow standard Lua-isms, this should probably be passed around as an opaque object provided by the Lua library.

Reply:
Let's introduce some globals as we discussed off-GitHub.